src/sentry/sentry_metrics/querying/visitors/query_expression.py

```diff
@@ -290,78 +290,124 @@ class UnitsNormalizationVisitor(QueryExpressionVisitor[QueryExpression]):
+            if unit_family_and_unit is not None:
+                unit_family, reference_unit, unit = unit_family_and_unit
+                return WithUnit(
+                    unit_family=unit_family, reference_unit=reference_unit, unit=unit
+                ), unit.apply_on_query_expression(timeseries)
 
-            # We set the first seen unit family, irrespectively if a unit is found, since if it's not found, the family
-            # will be unknown.
-            self._unit_family = unit_family
+        return WithNoUnit(), timeseries
 
-            if reference_unit is not None and unit is not None:
-                self._reference_unit = reference_unit
-                self._scaling_factor = unit.scaling_factor
-                return unit.apply_on_timeseries(timeseries)
+    def _visit_int(self, int_number: float) -> tuple[UnitMetadata, QueryExpression]:
+        return WithFutureUnit(), int_number
 
-        return timeseries
+    def _visit_float(self, float_number: float) -> tuple[UnitMetadata, QueryExpression]:
+        return WithFutureUnit(), float_number
+
+    def _visit_string(self, string: str) -> tuple[UnitMetadata, QueryExpression]:
+        return WithNoUnit(), string
 
     def _extract_unit(self, timeseries: Timeseries) -> str | None:
-        # If the aggregate doesn't support unit normalization, we will skip it.
         if timeseries.aggregate in self.UNITLESS_AGGREGATES:
-            raise NonNormalizableUnitsError(
-                f"The aggregate {timeseries.aggregate} doesn't need unit normalization"
-            )
+            return None
 
         parsed_mri = parse_mri(timeseries.metric.mri)
         if parsed_mri is not None:
             return parsed_mri.unit
 
-        raise NonNormalizableUnitsError(
-            "Units normalization can't be run if not all components have a metric mri"
-        )
+        return None
 
-    def get_units_metadata(
-        self,
-    ) -> tuple[UnitFamily | None, MeasurementUnit | None, float | int | None]:
+    def _is_numeric_scalar(self, value: QueryExpression) -> bool:
+        return isinstance(value, int) or isinstance(value, float)
+
+    def _normalize_future_units(self, unit: Unit, value: QueryExpression) -> QueryExpression:
         """
-        Returns metadata of the units that were encountered during the traversal.
+        Normalizes all future units, which in our case are just numeric scalars, using a common unit. This assumes
+        that such numbers are used in the context of the unit and as such they need to be scaled by a certain factor
+        to be normalized to the reference unit.
         """
-        # If we have a formula, we do not return the scaling factor, since a formula technically has multiple scaling
-        # factors, but they won't be of use to the frontend.
-        if self._is_formula:
-            return self._unit_family, self._reference_unit, None
+        return NumericScalarsNormalizationVisitor(unit).visit(value)
```

Comment:
We can naively use this visitor to normalize all scalars in the subtree, since we know that this will be called only once on a subtree, because once this is called, the entire subtree will be normalized. Any operand that will be added in the future, will have its own subtree.